### PR TITLE
Add element class to the /source route

### PR DESCRIPTION
**Pull request by qmfrederik**
Adds support for retrieving the underlying element class (such as `UIView` and its derivates) and adding it to the `source`.

### CODE REVIEW

**Comment by qmfrederik** [comment]
@marekcirkos I'm mainly creating the PR to get your feedback at this time. I think getting the class name can be of interest. It's hidden in the additionalAttributes dictionary and the value of 5004 seems to be a constant.

If you want to land this, let me know, and I'd be happy to work with you to get this PR ready to merge.

**Comment by mykola-mokhnach** [comment]
I'd rather add this property info FBElement itself. Also, it would be interesting to know if this magic hack works for all snapshot or it can potentially throw an exception. Is there any documentation about it?

**Comment by marekcirkos** [comment]
Don't we have this info already? All name in inspector come from classes?

**Comment by qmfrederik** [comment]
@marekcirkos Let me double check, until now I only found XCUI Test-related classes, not UIKit classes (such as `UIView` and friends).

**Comment by qmfrederik** [comment]
@marekcirkos Take for example the Facebook login page. For the text field with the user e-mail address, you now get:

```
<XCUIElementTypeTextField type="XCUIElementTypeTextField" value="Email address or phone number" name="username-field" label="Username Field" enabled="true" visible="true" x="28" y="302" width="324" height="44"/>
```

This PR would change that to:

```
<XCUIElementTypeTextField type="XCUIElementTypeTextField" value="Email address or phone number" name="username-field" class="UITextField" label="Username Field" enabled="true" visible="true" x="28" y="302" width="324" height="44"/>
```

so adding `class="UITextField"`. I checked the inspector but I don't think it is based on these classes - rather, it appears to use the XCUIElementTypes?

**Comment by qmfrederik** [comment]
@mykola-mokhnach No, I don't think there's a lot of documentation on this; you can see the various values in the debugger; I also found amex using it: [snapshot_summary.rb](https://github.com/americanexpress/xcode-result-bundle-processor/blob/master/lib/xcoderesultbundleprocessor/snapshot_summary.rb)

**Comment by mykola-mokhnach** [comment]
I checked the Ruby code and there it is verified (probably, because of a reason) whether an element provides the corresponding attribute first before getting the value.

```ruby
if element_snapshot.key?(:additionalAttributes) and element_snapshot[:additionalAttributes].key?(5004.to_s.to_sym)
        element_class = element_snapshot[:additionalAttributes][5004.to_s.to_sym]
      end
```

This means the property might not be consistent to all UI elements - some of them may have it and some other may not.

**Comment by qmfrederik** [comment]
Hmm, a blame on that file shows that code is 2 years old and there's not much info in the logs.

@mykola-mokhnach What's your take on it? Worth adding, even if it may not be present for all elements?

**Comment by mykola-mokhnach** [comment]
This property probably shows the old class name used by UIAutomation. If this is really like this then I saw some cases when UIAutomation class name was more specific than XCTest's one (which is always **Other** for "unknown" elements). Although I'm not sure about whether one can rely on it, since Apple didn't document it anywhere. Apple might remove the property one day after iOS update and then all automation clients will have to change their locators.

**Comment by qmfrederik** [comment]
@marekcirkos @mykola-mokhnach I updated the PR so that the code which fetches the class, is now part of FBElement. The main purpose is to get some additional information which can be used for debugging.

For example, for the status bar you would get values such as:
```
<XCUIElementTypeOther type="XCUIElementTypeOther" name="Bluetooth on" class="UIStatusBarBluetoothItemView" label="Bluetooth on" enabled="true" visible="true" x="323" y="0" width="8" height="20"/>
```

i.e. the `class="UIStatusBarBluetoothItemView"` communicates some information about the element which you cannot get from the `XCUIElementTypeOther` element type.

Let me know if you want to proceed with this.

File: WebDriverAgentLib/Categories/XCUIElement+FBClass.m [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+@end
+
+@implementation XCElementSnapshot (FBClass)
+
+- (NSString*)fb_class
+{
+  NSNumber* classId = [NSNumber numberWithInt:5004];
+  NSObject* class = [self.additionalAttributes objectForKey:classId];
+  if([class isKindOfClass:[NSString class]]){
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClass.m]: check if **class** is not equal to nil first

File: WebDriverAgentLib/Categories/XCUIElement+FBClass.m [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    return (NSString*)class;
+  } else {
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClass.m]: _else_ is not needed here

File: WebDriverAgentLib/Categories/XCUIElement+FBClass.m [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+#import "XCUIElement+FBClass.h"
+
+@implementation XCUIElement (FBClass)
+
+- (NSString*)fb_class
+{
+  if (!self.lastSnapshot) {
+    [self resolve];
+  }
+  return [self.lastSnapshot fb_class];
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClass.m]: there is fb_lastSnapshot method in Utilities category, which resolves the element automatically

File: WebDriverAgentLib/Categories/XCUIElement+FBClass.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+#import <WebDriverAgentLib/XCElementSnapshot.h>
+#import <WebDriverAgentLib/XCUIElement.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface XCUIElement (FBClass)
+
+/*! Element's class (UIView or derived) */
+@property (nonatomic, readonly, copy) NSString *fb_class;
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClass.h]: is it nullable or not?

File: WebDriverAgentLib/Categories/XCUIElement+FBWebDriverAttributes.m [hunk 5]
```diff
@@ -90,6 +91,11 @@ - (NSString *)wdType
   return [FBElementTypeTransformer stringWithElementType:self.elementType];
 }
 
+- (NSString *)wdClass
+{
+  return FBTransferEmptyStringToNil(self.fb_class);
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBWebDriverAttributes.m]: what the point to transform it to nil here if this can be already done in the corresponding extension method?

File: WebDriverAgentLib/Utilities/FBXPath.m [hunk 6]
```diff
@@ -224,6 +224,13 @@ + (int)recordElementAttributes:(xmlTextWriterPtr)writer forElement:(XCElementSna
       return rc;
     }
   }
+  if (element.wdClass) {
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBXPath.m]: this is going to break several unit and integration tests. Please check and update these

File: WebDriverAgentLib/Categories/XCUIElement+FBClass.m (hunk 1 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClass.m]:
```objc
id class = [self.additionalAttributes objectForKey:classId];
if (class && [class isKindOfClass:[NSString class]] && [(NSString *)class length]) {
  return (NSString *)class;
}
return nil;
```

Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClass.m]: If it is `nil` it will resolve to false anyway

File: WebDriverAgentLib/Categories/XCUIElement+FBClass.m [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+@implementation XCElementSnapshot (FBClass)
+
+- (NSString*)fb_class
+{
+  NSNumber* classId = [NSNumber numberWithInt:5004];
+  NSObject* class = [self.additionalAttributes objectForKey:classId];
+  if([class isKindOfClass:[NSString class]]){
+    return (NSString*)class;
+  } else {
+    return @"";
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClass.m]: Why not use `nil` here?

File: WebDriverAgentLib/Categories/XCUIElement+FBClass.m [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+@end
+
+@implementation XCElementSnapshot (FBClass)
+
+- (NSString*)fb_class
+{
+  NSNumber* classId = [NSNumber numberWithInt:5004];
+  NSObject* class = [self.additionalAttributes objectForKey:classId];
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClass.m]: Please use `id` instead

File: WebDriverAgentLib/Categories/XCUIElement+FBClass.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+@end
+
+
+@interface XCElementSnapshot (FBClass)
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClass.h]: Probably `FBUIClassName` would be better

**Comment by marekcirkos** [comment]
I am bit resistan to this diff in general. What is the gain here? You are duplicating `XCT` class names here. Plus client shouldn't care much about classes anyway.

@mykola-mokhnach any thoughts on this?

**Comment by mykola-mokhnach** [comment]
@marekcirkos this can sometimes help to locate an element if it has no accessibility identifier and XCTest recognises it as ElementTypeOther. This is the only possible advantage there. Also, I'm not sure about the actual performance. @qmfrederik do you see any visible perf degradation while getting the same source with and without class property?

**Comment by mykola-mokhnach** [comment]
Altogether I tend to agree with @marekcirkos 
There is no point to add more attributes unless they are really necessary.

**Comment by qmfrederik** [comment]
@mykola-mokhnach @marekcirkos 

Thanks for the feedback!

As to the "why": yes, the idea is to get additional information for elements that come back as "XCUIElementTypeOther", for example in the status bar you have:

```
<XCUIElementTypeOther type="XCUIElementTypeOther" class="UIStatusBarBluetoothItemView" />
```

So it's not merely duplicating; it's adding additional information.

As to the performance: I measured it; it's essentially reading a value from a dictionary (and that dict was prepopulated); the cost (on an iPhone 7) is in the range of nanoseconds per call; so even if you have say 100 elements it will only add 1 ms to the output.

As to unit tests: certainly, I'll add those!

**Comment by marekcirkos** [comment]
@qmfrederik I really appreciate your support here. However I can't see the point of serving this information. 
- You wouldn't be able to search elements for those classes or it will be really slow as it will need to serialise every element. Main reason why xpath is slow. Probably you want to use `xpath` anyway?
- It will be confusing which class name should be used to locate element with `by class name` search.

**Comment by qmfrederik** [comment]
@marekcirkos The main use case here is people who are trying to "dynamically" automate apps - that is, they have certain algorithms, such as "try to log on" which will try to find the user name + password text boxes and the login button and then enter credentials and try to log in.

In those cases, they heavily rely on /source (hence also the other PR related to visible) to get a full dump of the UI hierarchy, execute an algorithm to identify the relevant elements, and then execute actions against it.

So basically their requirement is to have as much information in /source as possible. It's fine if not all that information can be used in XPath or other queries - they can work around it.

I'm also OK with you not taking this PR - we try to contribute back our modifications to upstream open source projects, but it's up to you to define what you want to take and what not :)

**Comment by marekcirkos** [comment]
I wonder if it would make sense to drop XCTest class names in favour of iOS class names.

@mykola-mokhnach any thoughts on this?

**Comment by mykola-mokhnach** [comment]
And break thousands of existing automation scripts? No thanks :)

**Comment by marekcirkos** [comment]
I am sorry we will have to pass on this.
